### PR TITLE
Make conf_diagnostics apply also to the SSL conf errors

### PR DESCRIPTION
**Pull request by t8m**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
This is a draft PR - the CI failures demonstrate that conf_diagnostics work and make SSL_CTX_new to fail.

**Comment by t-j-h** [comment]
Looks good to me in terms of improving what we currently have ...

**Comment by beldmit** [comment]
I'd like to have some syntax check for the config file as a part of the openssl command line util but it can be a separate PR

File: crypto/context.c [hunk 1]
```diff
@@ -666,3 +667,19 @@ CRYPTO_THREAD_LOCAL *ossl_lib_ctx_get_rcukey(OSSL_LIB_CTX *libctx)
...
+        return 0;
+    return libctx->conf_diagnostics;
+}
+
+void OSSL_LIB_CTX_set_conf_diagnostics(OSSL_LIB_CTX *libctx, unsigned int value)
+{
+    libctx = ossl_lib_ctx_get_concrete(libctx);
+    if (libctx == NULL)
+        return;
+    libctx->conf_diagnostics = value != 0;
```
Comment (paulidale) [inline crypto/context.c:684]: Does this need to be locked/atomic?  Or can we guarantee that it isn't sensible to access in a threading situation?

It's a bit field so every access could require multiple instructions.  At least, this is true when writing, reading is less obvious.  Alternatively, make it not a bit field and atomic.

Comment (t8m) [reply crypto/context.c:684]: I do not think this particular value needs to be locked but there is a good point about the bit field. I'll change that.

File: ssl/ssl_mcnf.c [hunk 2]
```diff
@@ -60,6 +64,7 @@ static int ssl_do_config(SSL *s, SSL_CTX *ctx, const char *name, int system)
         SSL_CONF_CTX_set_ssl_ctx(cctx, ctx);
         libctx = ctx->libctx;
     }
+    conf_diagnostics = OSSL_LIB_CTX_get_conf_diagnostics(libctx);
```
Comment (mattcaswell) [inline ssl/ssl_mcnf.c:67]: if `conf_diagnostics` is true should we set the `SSL_CONF_FLAG_SHOW_ERRORS` flag on `cctx`?

File: crypto/context.c [hunk 3]
```diff
@@ -666,3 +667,19 @@ CRYPTO_THREAD_LOCAL *ossl_lib_ctx_get_rcukey(OSSL_LIB_CTX *libctx)
         return NULL;
     return &libctx->rcu_local_key;
 }
+
+int OSSL_LIB_CTX_get_conf_diagnostics(OSSL_LIB_CTX *libctx)
```
Comment (mattcaswell) [inline crypto/context.c:671]: Obviously these new functions will need documentation - and we should have a test for them.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/context.c (hunk 3 above)
Comment (t8m) [reply crypto/context.c:671]: Yes, sure.

File: ssl/ssl_mcnf.c (hunk 2 above)
Comment (t8m) [reply ssl/ssl_mcnf.c:67]: Yeah, good point.

**Comment by mattcaswell** [comment]
The CI failure looks relevant

File: CHANGES.md [hunk 4]
```diff
@@ -39,7 +39,13 @@ OpenSSL 3.4
  * The X25519 and X448 key exchange implementation in the FIPS provider
    is unapproved and has `fips=no` property.
 
-   * Tomas Mraz*
+   *Tomáš Mráz*
+
+ * Setting `config_diagnostics=1` in config file will cause returning
+   errors from SSL_CTX_new() and SSL_CTX_new_ex() if there is an error
```
Comment (mattcaswell) [inline CHANGES.md:45]: Setting `config_diagnostics=1` in  the config file will cause errors to be returned from SSL_CTX_new() ...

File: doc/man3/OSSL_LIB_CTX_set_conf_diagnostics.pod [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+=head1 SYNOPSIS
+
+ #include <openssl/crypto.h>
+
+ void OSSL_LIB_CTX_set_conf_diagnostics(OSSL_LIB_CTX *ctx, int value);
+ int OSSL_LIB_CTX_get_conf_diagnostics(OSSL_LIB_CTX *ctx);
+
+=head1 DESCRIPTION
+
+OSSL_LIB_CTX_set_conf_diagnostics() sets value of the configuration
```
Comment (mattcaswell) [inline doc/man3/OSSL_LIB_CTX_set_conf_diagnostics.pod:17]: sets the value

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by t8m** [comment]
> The CI failure looks relevant

The question is whether the failing test really makes sense. It basically loads a config into a single libctx concurrently from multiple threads. What is the semantics of such call? If the same file is being loaded than it could somewhat work, although I would not be surprised if there are some strange circumstances where it would not. However if a different file would be loaded concurrently into the same libctx, what the application can expect then? Random results.

I am very much inclined to drop the testcase.

**Review by paulidale — APPROVED** [review]

File: include/openssl/crypto.h.in [hunk 6]
```diff
@@ -537,7 +537,7 @@ void OSSL_LIB_CTX_free(OSSL_LIB_CTX *);
 OSSL_LIB_CTX *OSSL_LIB_CTX_get0_global_default(void);
 OSSL_LIB_CTX *OSSL_LIB_CTX_set0_default(OSSL_LIB_CTX *libctx);
 int OSSL_LIB_CTX_get_conf_diagnostics(OSSL_LIB_CTX *ctx);
-void OSSL_LIB_CTX_set_conf_diagnostics(OSSL_LIB_CTX *ctx, unsigned int value);
+void OSSL_LIB_CTX_set_conf_diagnostics(OSSL_LIB_CTX *ctx, int value);
```
Comment (nhorman) [inline include/openssl/crypto.h.in:540]: Is this considered ABI breakage?

**Review by nhorman — APPROVED** [review]

**Comment by t8m** [comment]
I've merged this yesterday. Thank you for the reviews.

**Comment by xnox** [comment]
```console
make test TESTS=test_sysdefault
...
make[2]: Entering directory '/home/xnox/upstream/openssl'
( SRCTOP=. \
  BLDTOP=. \
  PERL="/usr/bin/perl" \
  FIPSKEY="f4556650ac31d35461610bac4ed81b1a181b2d8a43ea2854cbae22ca74560813" \
  EXE_EXT= \
  /usr/bin/perl ./test/run_tests.pl test_sysdefault )
00-prep_fipsmodule_cnf.t .. ok   
All tests successful.
Files=1, Tests=1,  0 wallclock secs ( 0.01 usr  0.00 sys +  0.27 cusr  0.05 csys =  0.33 CPU)
Result: PASS
90-test_sysdefault.t .. 
    # ERROR: (ptr) 'ctx = SSL_CTX_new(TLS_method()) != NULL' failed @ test/sysdefaulttest.c:25
    # 0x0
    # 80CB81B9097E0000:error:0A0001A3:SSL routines:SSL_CTX_new_ex:error in system default config:ssl/ssl_lib.c:4100:
    # OPENSSL_TEST_RAND_SEED=1715381166
    not ok 1 - test_func
# ------------------------------------------------------------------------------
../../util/wrap.pl ../../test/sysdefaulttest => 1
not ok 1 - sysdefaulttest
90-test_sysdefault.t .. 1/? ----------------------------------------------------
#   Failed test 'sysdefaulttest'
#   at test/recipes/90-test_sysdefault.t line 23.
90-test_sysdefault.t .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 

Test Summary Report
-------------------
90-test_sysdefault.t (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=1, Tests=1,  0 wallclock secs ( 0.01 usr  0.01 sys +  0.17 cusr  0.04 csys =  0.23 CPU)
Result: FAIL
make[2]: *** [Makefile:3949: run_tests] Error 1
make[2]: Leaving directory '/home/xnox/upstream/openssl'
make[1]: *** [Makefile:3946: _tests] Error 2
make[1]: Leaving directory '/home/xnox/upstream/openssl'
make: *** [Makefile:3944: tests] Error 2
```

Fails for me, and git bisect blames:

```
21819f78b057c254254646a7854bfad0cd40ed83 is the first bad commit
commit 21819f78b057c254254646a7854bfad0cd40ed83
Date:   Fri Apr 26 17:23:13 2024 +0200

    Make conf_diagnostics apply also to the SSL conf errors
    
    Reviewed-by: Paul Dale <ppzgs1@>
    Reviewed-by: Neil Horman <nhorman@openssl.org>
    (Merged from https://github.com/openssl/openssl/pull/24275)

 crypto/conf/conf_mod.c      |  1 +
 crypto/context.c            | 17 +++++++++++++++++
 crypto/err/openssl.txt      |  1 +
 include/openssl/crypto.h.in |  2 ++
 include/openssl/sslerr.h    |  1 +
 ssl/ssl_err.c               |  2 ++
 ssl/ssl_lib.c               |  5 ++++-
 ssl/ssl_local.h             |  2 +-
 ssl/ssl_mcnf.c              | 13 +++++++++----
 test/sysdefaulttest.c       | 18 ++++++++----------
 util/libcrypto.num          |  2 ++
 11 files changed, 48 insertions(+), 16 deletions(-)
```

This is on Ubuntu 24.40 host x86-64 with gcc-13.

**Comment by t8m** [comment]
@xnox please open a new issue for this. Is this a clean build without any patches on top?

**Pull request closed without merging**